### PR TITLE
TRCA variation

### PR DESCRIPTION
**Pull request by ludovicdmt**
Hello, 

In addition to the original method of TRCA, I propose in this PR a variation that use regularization and riemannian mean for the computation of the matrix S (instead of euclidean mean as in original implementation). It provides improvements when data are noisy and have access to large number of calibration data.   

This variation is inspired by a similar work from A. Barachant on CSP: https://hal.archives-ouvertes.fr/hal-00602686/document  

Ideas are :
- Empirical estimator for covariance matrix used in original paper is unbiased but with high variance especially with a high number of channel. It could benefit of some regularization.
- The matrix S in TRCA method represent the average (euclidean) inter-trial covariance matrix. Covariance matrices are semi-definite positive and lie on a riemannian manifold. A measure of distance (geodesic distance) is defined on this curvated manifold: the riemannian mean. Euclidean mean does not respect the curvature of the space.   
However the quality of the estimation of the riemannian mean depends on the number of covariance matrices used.  

It does not provide an improvement of performance for the specific dataset used in ./testd/data/trcadata.mat but we notice at least 10% of improvement on our data that are more noisy and use more calibration data (about 10 trial per class).  

This variation rely on Pyriemann toolbox which is already in requirements.

### CODE REVIEW

File: meegkit/trca.py [hunk 1]
```diff
@@ -71,6 +74,74 @@ def trca(X):
     return W_best
 
 
+def trca_regul(X, regul):
+    """Task-related component analysis.
+
+    This function implements a variation of the method described in [1].
```
Comment (nbara) [inline meegkit/trca.py:80]:
```suggestion
    This function implements a variation of the method described in [1]_.
```

Can you also reference the Barachant document here?

File: meegkit/trca.py [hunk 2]
```diff
@@ -149,7 +234,14 @@ def fit(self, X, y):
                     trains[class_i, fb_i] = eeg_tmp
                 # Find the spatial filter for the corresponding filtered signal
                 # and label
-                w_best = trca(eeg_tmp)
+                if self.method == 'original':
+                    w_best = trca(eeg_tmp)
+                elif self.method == 'riemann':
+                    w_best = trca_regul(eeg_tmp, self.regul)
```
Comment (nbara) [inline meegkit/trca.py:240]: It would be great if the riemannian method was called at least once in the test. 

```python
@pytest.mark.parametrize('ensemble', [True, False])
@pytest.mark.parametrize('method', ['original', 'riemann'])
def test_trcacode(ensemble):
```

See this file:
https://github.com/nbara/python-meegkit/blob/4aa4ba49354b996be20eda41660a550d1bd31f9a/tests/test_trca.py#L45

File: meegkit/utils/trca.py [hunk 3]
```diff
@@ -136,3 +136,46 @@ def bandpass(eeg, sfreq, Wp, Ws):
...
+            \hat{\Sigma} = (1 - \gamma)\Sigma_{scm} + \gamma T
+    where :math:`T` is the diagonal target matrix:
+    .. math::
+            T_{i,j} = \{ \Sigma_{scm}^{ii} \text{if} i = j,
+            0 \text{otherwise} \}
+    Note that the optimal :math:`\gamma` is estimate by the authors' method.
+
+    Parameters
+    ----------
+    X: Signal matrix, Nchannels X Nsamples
```
Comment (nbara) [inline meegkit/utils/trca.py:155]:
```suggestion
    X: array, shape=(n_channels, n_samples) 
        Signal matrix.
```

here and everywhere

File: meegkit/utils/trca.py [hunk 4]
```diff
@@ -136,3 +136,46 @@ def bandpass(eeg, sfreq, Wp, Ws):
     y = filtfilt(B, A, eeg, axis=0, padtype='odd',
                  padlen=3 * (max(len(B), len(A)) - 1))
     return y
+
+
+def schaefer_strimmer_cov(X):
+    r"""Schaefer-Strimmer covariance estimator.
+
+    Shrinkage estimator using method from [1]:
```
Comment (nbara) [inline meegkit/utils/trca.py:144]:
```suggestion
    Shrinkage estimator using method from [1]_:
    
```

File: meegkit/utils/trca.py [hunk 5]
```diff
@@ -136,3 +136,46 @@ def bandpass(eeg, sfreq, Wp, Ws):
...
+    .. math::
+            \hat{\Sigma} = (1 - \gamma)\Sigma_{scm} + \gamma T
+    where :math:`T` is the diagonal target matrix:
```
Comment (nbara) [inline meegkit/utils/trca.py:147]: I think we need an empty line between semicolon and the formula for it to render. 

Also, there are a few errors in the docstring: 

```
/home/travis/build/nbara/python-meegkit/meegkit/utils/trca.py:docstring of meegkit.utils.trca.schaefer_strimmer_cov:6: WARNING: Unexpected indentation.

/home/travis/build/nbara/python-meegkit/meegkit/utils/trca.py:docstring of meegkit.utils.trca.schaefer_strimmer_cov:7: WARNING: Literal block ends without a blank line; unexpected unindent.

/home/travis/build/nbara/python-meegkit/meegkit/utils/trca.py:docstring of meegkit.utils.trca.schaefer_strimmer_cov:9: WARNING: Unexpected indentation.

/home/travis/build/nbara/python-meegkit/meegkit/utils/trca.py:docstring of meegkit.utils.trca.schaefer_strimmer_cov:11: WARNING: Literal block ends without a blank line; unexpected unindent.
```

https://travis-ci.com/github/nbara/python-meegkit/builds/223607667#L621

File: meegkit/utils/trca.py [hunk 6]
```diff
@@ -136,3 +136,46 @@ def bandpass(eeg, sfreq, Wp, Ws):
...
+            T_{i,j} = \{ \Sigma_{scm}^{ii} \text{if} i = j,
+            0 \text{otherwise} \}
+    Note that the optimal :math:`\gamma` is estimate by the authors' method.
+
+    Parameters
+    ----------
+    X: Signal matrix, Nchannels X Nsamples
+    Returns
+    -------
+    cov: Schaefer-Strimmer shrinkage covariance matrix, Nchannels X Nchannels
```
Comment (nbara) [inline meegkit/utils/trca.py:158]:
```suggestion
    cov: array, shape=(n_channels, n_channels)
        Schaefer-Strimmer shrinkage covariance matrix.
```

**Review by nbara — COMMENTED** [review]
Hi @ludovicdmt, that's great thanks a lot! I've left a few comments if you could please address them? 

If you don't have time I'll do it myself later this week. I'll also try to improve the example_trca.py a bit

**Comment by ludovicdmt** [comment]
Thanks for the feedback, I'll have a look and do that tomorrow.

**Comment by ludovicdmt** [comment]
Should be good now

**Comment by nbara** [comment]
I'm a bit concerned by the very poor performance compared to the original method. 

A performance drop can be expected, but in this cas we sometimes get ~30% accuracy on the example dataset

File: meegkit/trca.py [hunk 7]
```diff
@@ -210,9 +179,154 @@ def predict(self, X):
...
+    UX -= np.mean(UX, 1)[:, None]
+
+    # Compute empirical variance of all data (to be bounded)
+    cov = Covariances(estimator=method).fit_transform(UX[np.newaxis, ...])
+    Q = np.squeeze(cov)
+
+    # 2. Compute average empirical covariance between all pairs of trials
+    # -------------------------------------------------------------------------
+    # Intertrial correlation computation
+    data = np.concatenate((X, X), axis=1)
```
Comment (nbara) [inline meegkit/trca.py:309]: should the data not me zero-centered here before computing the covariance?

Comment (ludovicdmt) [reply meegkit/trca.py:309]: Actually, `pyriemann.estimation.Covariances` [rely](https://github.com/alexandrebarachant/pyRiemann/blob/master/pyriemann/utils/covariance.py) on `sklearn.covariance` and by default `assume_centered` is set to `False` (in `schaefer_strimmer_cov` function the data are zero-centered as well).

**Comment by ludovicdmt** [comment]
Regarding performance, as explained in [1](https://hal.archives-ouvertes.fr/hal-01570120/document) geodesic mean is more robust to outliers than the euclidean mean. But it is also really sensible to ill-conditioned covariance matrices.    
In this dataset, data are especially clean (an accuracy of 100% on some blocks). In addition, they are only 4 samples per class to estimate the $6\times6$ S matrix, so probably ill-conditioned estimations of covariance matrix S.   

On data with smaller SNR and with more training samples, this TRCA variation gave us ~10% of performance improvement. I think the two methods are complementary in terms of use cases.   

But maybe that's not the point of Meegkit toolbox which is more about collecting _standard_ methods and I can't definitely understand that.

**Comment by nbara** [comment]
The fact that the riemann method doesn't work as well on this dataset is a problem in itself, it's just that it makes it difficult to unit-test. For the original implementation, I can be reasonably sure that my code is correct because it yields good classification results. Here it's a bit more complex, and it will make catching potential future bugs (due to dependency changes, etc.) more difficult to track

**Comment by nbara** [comment]
OK I restricted the test scope, and added an illustration to the example. Thanks @ludovicdmt

**Comment by ludovicdmt** [comment]
Ok thank you for the help and work @nbara !
